physics/module_bl_mynn.F90

```diff
@@ -480,146 +444,166 @@ SUBROUTINE mynn_bl_driver(            &
+    real(kind_phys), DIMENSION(:,:), INTENT(in) :: dz,       &
          &u,v,w,th,sqv3D,p,exner,rho,T3D
-    REAL, DIMENSION(:,:), INTENT(in):: &
-         &sqc3D,sqi3D,qni,qnc,qnwfa,qnifa
-    REAL, DIMENSION(:,:), INTENT(in):: ozone
-    REAL, DIMENSION(:), INTENT(in) :: xland,ust,       &
-         &ch,ts,qsfc,ps,hfx,qfx,wspd,uoce,voce,vdfg,znt
-
-    REAL, DIMENSION(:,:), INTENT(inout) ::       &
+    real(kind_phys), DIMENSION(:,:), INTENT(in) ::           &
+         &sqc3D,sqi3D,sqs3D,qni,qnc,qnwfa,qnifa,qnbca
+    real(kind_phys), DIMENSION(:,:), INTENT(in):: ozone
+    real(kind_phys), DIMENSION(:),   INTENT(in):: ust,       &
+         &ch,qsfc,ps,wspd
+    real(kind_phys), DIMENSION(:,:), INTENT(inout) ::        &
          &Qke,Tsq,Qsq,Cov,qke_adv
+    real(kind_phys), DIMENSION(:,:), INTENT(inout) ::        &
+         &rublten,rvblten,rthblten,rqvblten,rqcblten,        &
+         &rqiblten,rqsblten,rqniblten,rqncblten,             &
+         &rqnwfablten,rqnifablten,rqnbcablten
+    real(kind_phys), DIMENSION(:,:), INTENT(inout) :: dozone
+    real(kind_phys), DIMENSION(:,:), INTENT(in)    :: rthraten
 
-    REAL, DIMENSION(:,:), INTENT(inout) ::       &
-         &RUBLTEN,RVBLTEN,RTHBLTEN,RQVBLTEN,RQCBLTEN,        &
-         &RQIBLTEN,RQNIBLTEN,RQNCBLTEN,                      &
-         &RQNWFABLTEN,RQNIFABLTEN
-    REAL, DIMENSION(:,:), INTENT(inout) :: DOZONE
-
-    REAL, DIMENSION(:,:), INTENT(in)    :: RTHRATEN
-
-    REAL, DIMENSION(:,:), INTENT(out)   ::       &
-         &exch_h,exch_m
+    real(kind_phys), DIMENSION(:,:), INTENT(out)   :: exch_h,exch_m
+    real(kind_phys), DIMENSION(:),   INTENT(in)    :: xland, &
+         &ts,znt,hfx,qfx,uoce,voce
 
    !These 10 arrays are only allocated when bl_mynn_output > 0
-   REAL, DIMENSION(:,:), INTENT(inout) :: &
+   real(kind_phys), DIMENSION(:,:), INTENT(inout) ::         &
          & edmf_a,edmf_w,edmf_qt,edmf_thl,edmf_ent,edmf_qc,  &
          & sub_thl3D,sub_sqv3D,det_thl3D,det_sqv3D
 
-!   REAL, DIMENSION(IMS:IME,KMS:KME)   :: &
+!   real, DIMENSION(IMS:IME,KMS:KME)   :: &
 !         & edmf_a_dd,edmf_w_dd,edmf_qt_dd,edmf_thl_dd,edmf_ent_dd,edmf_qc_dd
 
-    REAL, DIMENSION(:), INTENT(inout) :: Pblh,rmol
+    real(kind_phys), DIMENSION(:), INTENT(inout) :: Pblh
+    real(kind_phys), DIMENSION(:), INTENT(inout) :: rmol
 
-    REAL, DIMENSION(IMS:IME) :: Psig_bl,Psig_shcu
+    real(kind_phys), DIMENSION(IMS:IME) :: psig_bl,psig_shcu
 
-    INTEGER,DIMENSION(:),INTENT(INOUT) ::             &
+    INTEGER,DIMENSION(:),INTENT(INOUT) ::                    &
          &KPBL,nupdraft,ktop_plume
 
-    REAL, DIMENSION(:), INTENT(OUT) ::                &
-         &maxmf
+    real(kind_phys), DIMENSION(:), INTENT(out) ::  maxmf
 
-    REAL, DIMENSION(:,:), INTENT(inout) ::      &
-         &el_pbl
+    real(kind_phys), DIMENSION(:,:), INTENT(inout) :: el_pbl
 
-    REAL, DIMENSION(:,:), INTENT(out) ::          &
+    real(kind_phys), DIMENSION(:,:), INTENT(out) ::          &
```

Comment:
All of these are only conditionally given values (`if tke_budget`). Since they're intent(out), they will be filled with garbage (rather than 0 or something else) when `tke_budget == .false.`. It probably doesn't matter since they're diagnostic, but it is one of the CCPP rules that intent(out) variables are initialized. Even though they are intent(inout) in the wrapper, they will be overwritten by garbage in here when `tke_budget == .false.`.

Reply:
I think I'll change them to inout, since in a more free world, other schemes may be allowed to modify the TKE budget.